### ISSUE TITLE
NoSuchMethodError trying to iterate ConcurrentHashMap on Android <= 9

### ISSUE THREAD

**Issue by mattjohnsonpint** [xamarin/xamarin-android#7150]
In a `net6.0-android` project referencing a Java jar using `AndroidLibrary` that exposes a `java.util.concurrent.ConcurrentHashMap`, trying to iterate over the items (or keys) from .NET will work on Android 10 and higher.  But on Android 9 (and presumably lower) it throws the following:

```
java.lang.NoSuchMethodError: no non-static method "Ljava/util/concurrent/ConcurrentHashMap$KeyIterator;.hasNext()Z"
    at crc649e6e32e78efd8fae.MainActivity.n_onCreate(Native Method)
    at crc649e6e32e78efd8fae.MainActivity.onCreate(MainActivity.java:29)
    at android.app.Activity.performCreate(Activity.java:7136)
    at android.app.Activity.performCreate(Activity.java:7127)
    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2893)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
    at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:193)
    at android.app.ActivityThread.main(ActivityThread.java:6669)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

Java code:
```java
package mylib;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class MyObject {

    private Map<String, String> data = new ConcurrentHashMap<>();

    public Map<String, String> getData() {
        return data;
    }

    public void setData(String key, String value) {
        data.put(key, value);
    }

    // ...
}
```

.NET Code:
```csharp
var myObject = new MyLib.MyObject();
myObject.SetData("a", "1");
myObject.SetData("b", "2");
myObject.SetData("c", "3");

var data = myObject.Data!;
foreach (var item in data)    // throws on iteration
{
    Console.WriteLine($"{item.Key} = {item.Value}");
}
```

Reproduction:

- Clone https://github.com/mattjohnsonpint/zzz-XamarinAndroidConcurrentHashMapIssue
- Examine `mylib/MyObject.java` and `MyApp/MainActivity.cs` to see relevant code
- Call `./build.sh` - it should build the java project, then the C# project.  (adjust the script if necessary)
- Start an emulator (or use a real device) running Android 10 (API 29) or higher
- Call `./run.sh` - observe the app starts without issue
- Shut down that emulator and start another one running Android 9 (API 28)
- Call `./run.sh` - observe the app starts and immediately crashes
- Run `adb logcat | grep MonoDroid` (or similar) to see the stack trace, as posted above.

The only workaround I could figure out was to copy the `ConcurrentHashMap` values to a `HashMap` and iterate over that instead.  Surely there is a better way, but I couldn't figure it out.

```csharp
if (Build.VERSION.SdkInt <= BuildVersionCodes.P)
{
    // Workaround by copying the ConcurrentHashMap to a HashMap.
    // Inefficient, but works if you only need one-time access to read the items.
    var map = new HashMap((IDictionary) data);
    data = new JavaDictionary<string, string>(map.Handle, JniHandleOwnership.DoNotRegister);
}
```

I'm not sure why it throws `NoSuchMethodError`.  I examined the Android 9 source code and found `hasNext` [here](https://cs.android.com/android/platform/superproject/+/android-9.0.0_r1:libcore/ojluni/src/main/java/java/util/concurrent/ConcurrentHashMap.java;l=3452#:~:text=3451-,3452,-3453).  Maybe I'm looking in the wrong place?

**Comment by jpobst** [comment]
Does it work if you maintain the subtype instead of casting as `Map`?

```java
private ConcurrentHashMap<String, String> data = new ConcurrentHashMap<>();

public ConcurrentHashMap<String, String> getData() {
    return data;
}
```

**Comment by mattjohnsonpint** [comment]
If I define it that way, I get a C# compilation error trying to enumerate the items:

`error CS1579: foreach statement cannot operate on variables of type 'ConcurrentHashMap' because 'ConcurrentHashMap' does not contain a public instance or extension definition for 'GetEnumerator'`


If I take out the `foreach` and compile, I can see it comes through as `Java.Util.Concurrent.ConcurrentHashMap`. Trying a few things with that, with the following results:

- `data.KeySet()` still gives `Java.Lang.NoSuchMethodError: no non-static method "Ljava/util/concurrent/ConcurrentHashMap$KeyIterator;.hasNext()Z"`
- `data.EntrySet()` gives `Java.Lang.NoSuchMethodError: no non-static method "Ljava/util/concurrent/ConcurrentHashMap$EntryIterator;.hasNext()Z"`
- `Java.Util.Collections.List(data.Keys())` works
- `Java.Util.Collections.List(data.Elements())` works
- This works:
  ```csharp
  var keys = Java.Util.Collections.List(data.Keys());
  foreach (string key in keys)
  {
      var value = data.Get(new Java.Lang.String(key));
      Console.WriteLine($"{key} = {value}");
  }
  ```

- This works, but gives a warning:
  `[CA1416] This call site is reachable on: 'Android' 21.0 and later. 'ConcurrentHashMap.ForEach(IBiConsumer)' is only supported on: 'android' 24.0 and later.`

  ```csharp
  using var action = new MyAction();
  data.ForEach(action);

  // ...

  private class MyAction : Java.Lang.Object, IBiConsumer
  {
      public void Accept(Java.Lang.Object? key, Java.Lang.Object? value)
      {
          Console.WriteLine($"{key} = {value}");
      }
  }
```

I like the last one best, but wondering if there is an easier way to pass a C# delegate to `ForEach` method that expects `IBiConsumer`?

Also, since in my real app I can't change the implementation of the Java code, I have to do some fancy footwork in C#:
```csharp
var data = GetObject<ConcurrentHashMap>(((IJavaObject) myObject.Data!).Handle, JniHandleOwnership.DoNotTransfer)!;
```
A simple cast does not work at runtime.  Is that the right way?


Despite all of this for workarounds, it would be great if I didn't need any of them and the original code worked as expected.  It does on Android API 29 and higher. Can anything be done about that under the hood to make it "just work" on lower Android versions?

**Comment by jonpryor** [comment]
Apologies for the delay.  I'm still investigating, but in the meantime I have a smaller repro which doesn't involve the use of `MyObject.java`:

```csharp
using System.Collections.Generic;
using Android.Runtime;

namespace android_hw_net7;

[Activity(Label = "@string/app_name", MainLauncher = true)]
public class MainActivity : Activity
{
    protected override void OnCreate(Bundle? savedInstanceState)
    {
        base.OnCreate(savedInstanceState);

        // Set our view from the "main" layout resource
        SetContentView(Resource.Layout.activity_main);

        var cmap = new Java.Util.Concurrent.ConcurrentHashMap ();
        cmap.Put (new Java.Lang.String ("a"), new Java.Lang.String ("1"));
        cmap.Put (new Java.Lang.String ("b"), new Java.Lang.String ("2"));

        IDictionary<string, string> d =
            new Android.Runtime.JavaDictionary<string, string> (cmap.Handle, JniHandleOwnership.DoNotTransfer);
        foreach (var e in d) {
            Console.WriteLine ($"`{e.Key}` => `{e.Value}`");
        }
    }
}
```

**Comment by mattjohnsonpint** [comment]
No worries.  I'm working around it for now by copying the dictionary before iterating it - since I only need to read values once, and I'm at a point in my code where I'm not concerned with concurrency.

**Comment by jonpryor** [comment]
TL;DR: I think this is an Android 9 bug which was fixed in Android 10, and doesn't impact Android 8 or earlier.

Moreover, I can't think of anything that we could change on our end. :-(

---

I feel like I regularly say "I should read more closely", but…

I should read things more closely.

*Before* the crash, this "interesting" message is written to `adb logcat`:

```
Accessing hidden method Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasNext()Z (dark greylist, JNI)
```

In context:

```
Accessing hidden method Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasNext()Z (dark greylist, JNI)
Shutting down VM
FATAL EXCEPTION: main
Process: com.companyname.android_hw_net7, PID: 20526
java.lang.NoSuchMethodError: no non-static method "Ljava/util/concurrent/ConcurrentHashMap$KeyIterator;.hasNext()Z"
```

Which brings us to the [Restrictions on non-SDK interfaces](https://developer.android.com/guide/app-compatibility/restrictions-non-sdk-interfaces) Android docs, which links to the [Android 9 (API-28) `hiddenapi-light-greylist.txt` list](https://android.googlesource.com/platform/frameworks/base/+/pie-release/config/hiddenapi-light-greylist.txt) and the [Android 10 (API-29) `hiddenapi-flags.csv` file](https://dl.google.com/developers/android/qt/non-sdk/hiddenapi-flags.csv).

No "dark greylist" is listed for Android 9.  The Android 9 "light graylist" only contains the following entries which mention `ConcurrentHashMap`:

```
Landroid/media/AudioManager;->mAudioFocusIdListenerMap:Ljava/util/concurrent/ConcurrentHashMap;
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasMoreElements()Z
```

Compare and contrast with the Android 10 list (the format of which appears to have changed vs. Android 9):

```
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;-><init>([Ljava/util/concurrent/ConcurrentHashMap$Node;IIILjava/util/concurrent/ConcurrentHashMap;)V,greylist-max-o
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasMoreElements()Z,core-platform-api,public-api,system-api,test-api,whitelist
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasNext()Z,core-platform-api,public-api,system-api,test-api,whitelist
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->lastReturned:Ljava/util/concurrent/ConcurrentHashMap$Node;,greylist-max-o
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->map:Ljava/util/concurrent/ConcurrentHashMap;,greylist-max-o
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->remove()V,core-platform-api,public-api,system-api,test-api,whitelist
```

Plus *588 other entries*.

Further note the Android 10 entry for `BaseIterator.hasNext()`:

```
Ljava/util/concurrent/ConcurrentHashMap$BaseIterator;->hasNext()Z,core-platform-api,public-api,system-api,test-api,whitelist
```

The `hasNext()` method is *whitelisted* in Android 10.

(*Why*? Because it implements the `java.util.Iterator` interface, which is public!)

---

What appears to have happened -- and this is conjecture on my part -- is that when Google implemented Restrictions on non-SDK interfaces, their list was incomplete: it didn't include public members of non-public types which implemented public APIs.  Additionally, the restrictions are inconsistent: the docs imply that both Java Reflection & JNI usage will result in "similar" behavior -- an exception, a `NULL` return value -- but in *this* case I was able to use Java Reflection to invoke `ConcurrentHashMap.BaseIterator.hasNext()` without any exception being thrown.  (.NET bindings don't use Java Reflection, as JNI is faster.)

Consequently, Android 9 is "potentially bad" / fraught with peril, as any method which uses "abstraction" as intended -- use public abstract classes or interface types to hide private implementation details -- could potentially reference an "inadvertently greylisted" method.

I'm surprised we haven't hit this before!  (Just lucky, I guess?)

There is a "workaround", of sorts: setting [`/manifest/uses-sdk/@android:targetSdkVersion`](https://developer.android.com/guide/topics/manifest/uses-sdk-element#target) to "27" will allow the `hasNext` method to be executed.  However:

 1. When the app is launched, Android shows a dialog stating:
    > Detected problems with API compatibility.
    > (visit g.co/dev/appcompat for more info)

 2. The [Google Play store requires API-30 as the `targetSdkVersion`](https://developer.android.com/google/play/requirements/target-sdk).

Thus the workaround likely isn't viable.

How many customers would you lose if you set Android 10 as the minimum? ¯\\_(ツ)_/¯

**Comment by mattjohnsonpint** [comment]
> How many customers would you lose if you set Android 10 as the minimum?

I maintain a library that can be consumed by any .NET MAUI developer, and .NET MAUI has [declared support](https://docs.microsoft.com/en-us/dotnet/maui/supported-platforms) for Android 5.0 (API 21) or higher, so I set `SupportedOSPlatformVersion` to 21 in the csproj - just like the MAUI starter app template does.

I don't think it's worth raising the minimum supported version in my case.  I'll just keep using the workaround of copying the data before trying to read it.  Unless you can think of a cleaner workaround?

**Comment by mattjohnsonpint** [comment]
Thanks for the detailed investigation, BTW. Very insightful. 👍

**Comment by jonpryor** [comment]
After thinking about this further, there *is* something we can do to better support Android 9:

Fix: https://github.com/xamarin/java.interop/issues/910

For "reasons", when we have an instance of a Java side interface and wrap it into an "Invoker", method invocations are done against the runtime type of the Java interface.  Thus the "Invoker" for `Java.Util.IIterator.HasNext`:

```csharp
namespace Java.Util {

	internal partial class IIteratorInvoker : global::Java.Lang.Object, IIterator {
		IntPtr id_hasNext;
		public unsafe bool HasNext {
			get {
				if (id_hasNext == IntPtr.Zero)
					id_hasNext = JNIEnv.GetMethodID (class_ref, "hasNext", "()Z");
				return JNIEnv.CallBooleanMethod (((global::Java.Lang.Object) this).Handle, id_hasNext);
			}
		}
	}
}
```

This aligns with the managed stack trace that we get as part of the app crash:

```
Java.Lang.NoSuchMethodError: no non-static method "Ljava/util/concurrent/ConcurrentHashMap$KeyIterator;.hasNext()Z"
   at Java.Interop.JniEnvironment.InstanceMethods.GetMethodID(JniObjectReference type, String name, String signature)
   at Android.Runtime.JNIEnv.GetMethodID(IntPtr kls, String name, String signature)
   at Java.Util.IIteratorInvoker.get_HasNext()
   at android_hw_net7.MainActivity.OnCreate(Bundle savedInstanceState)
   at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState)
   at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0)
```

Specifically, the `JNIEnv.GetMethodID()` for `hasNext()Z` is done against the *runtime type* `java.util.concurrent.ConcurrentHashMap.KeyIterator`, and that is what fails.

*However*, that's not the only way we can implement interface invokers.  (This was the *easiest* way.)

The *other* way to implement interface invokers is described at https://github.com/xamarin/java.interop/issues/910, and when I use that style implementation, I'm able to invoke `Iterator.hasNext()` on the `java.util.concurrent.ConcurrentHashMap.KeyIterator` instance *without* prompting the "Accessing hidden method" message and runtime crash.

So…

Good News™, we can better support this scenario.  "Closed as dupe of xamarin/java.interop#910"

Bad News™, I'm not sure when we'll actually schedule fixing xamarin/java.interop#910…. Plus it's a `generator` fix, so *every binding assembly* would need to be rebuilt with this fix.  Wee?
